src/ECharts/Commands.purs

```diff
@@ -3,1115 +3,1128 @@ module ECharts.Commands where
-    xx = get "x" dsl
-    yy = get "y" dsl
-  in
-    set "coord" $ toForeign $ Arr.catMaybes [ xx, yy ]
-
-buildCenter ∷ ∀ i. DSL TP.PointI → DSL (center ∷ I|i)
-buildCenter dsl =
-  let
-    xx = get "x" dsl
-    yy = get "y" dsl
-  in
-    set "center" $ toForeign $ Arr.catMaybes [ xx, yy ]
-
-buildRadius ∷ ∀ i. DSL TP.RadiusI → DSL (radius ∷ I|i)
-buildRadius dsl =
-  let
-    s = get "start" dsl
-    e = get "end" dsl
-  in
-    set "radius" $ toForeign $ Arr.catMaybes [ s, e ]
-
-setStart ∷ ∀ i. DSL TP.DimensionI → DSL (start ∷ I|i)
-setStart dsl =
-  F.traverse_ (set "start") $ lastWithKeys ["pixels", "percents"] dsl
-
-setEnd ∷ ∀ i. DSL TP.DimensionI → DSL (end ∷ I|i)
-setEnd dsl =
-  F.traverse_ (set "end") $ lastWithKeys ["pixels", "percents"] dsl
-
-setBarWidth ∷ ∀ i. DSL TP.DimensionI → DSL (barWidth ∷ I|i)
-setBarWidth dsl =
-  F.traverse_ (set "barWidth") $ lastWithKeys ["pixels", "percents"] dsl
-
-setX ∷ ∀ i. DSL TP.DimensionI → DSL (x ∷ I|i)
-setX dsl =
-  F.traverse_ (set "x") $ lastWithKeys ["pixels", "percents"] dsl
-
-setY ∷ ∀ i. DSL TP.DimensionI → DSL (y ∷ I|i)
-setY dsl =
-  F.traverse_ (set "y") $ lastWithKeys ["pixels", "percents"] dsl
-
-setZ ∷ ∀ i. DSL TP.DimensionI → DSL (z ∷ I|i)
-setZ dsl =
-  F.traverse_ (set "z") $ lastWithKeys ["pixels", "percents"] dsl
-
-coordXIx ∷ ∀ i. Int → DSL (x ∷ I|i)
+buildCoord ∷ ∀ i m. Monad m ⇒ DSL TP.PointI m → DSL (coord ∷ I|i) m
+buildCoord dsl = do
+  xx ← get "x" dsl
+  yy ← get "y" dsl
+  set "coord" $ toForeign $ Arr.catMaybes [ xx, yy ]
+
+buildCenter ∷ ∀ i m. Monad m ⇒ DSL TP.PointI m → DSL (center ∷ I|i) m
+buildCenter dsl = do
+  xx ← get "x" dsl
+  yy ← get "y" dsl
+  set "center" $ toForeign $ Arr.catMaybes [ xx, yy ]
+
+buildRadius ∷ ∀ i m. Monad m ⇒ DSL TP.RadiusI m → DSL (radius ∷ I|i) m
+buildRadius dsl = do
+ s ← get "start" dsl
+ e ← get "end" dsl
+ set "radius" $ toForeign $ Arr.concat [s, e]
+
+setStart ∷ ∀ i m. Monad m ⇒ DSL TP.DimensionI m → DSL (start ∷ I|i) m
+setStart dsl = do
+  keys ← lastWithKeys ["pixels", "percents"] dsl
+  F.for_ (keys ∷ Array Foreign) $ set "start"
+
+setEnd ∷ ∀ i m. Monad m ⇒ DSL TP.DimensionI m → DSL (end ∷ I|i) m
+setEnd dsl = do
+  keys ← lastWithKeys ["pixels", "percents"] dsl
+  F.for_ (keys ∷ Array Foreign) $ set "end"
+
+setBarWidth ∷ ∀ i m. Monad m ⇒ DSL TP.DimensionI m → DSL (barWidth ∷ I|i) m
+setBarWidth dsl = do
+  keys ← lastWithKeys ["pixels", "percents"] dsl
+  F.for_ (keys ∷ Array Foreign) $ set "barWidth"
+
+setX ∷ ∀ i m. Monad m ⇒ DSL TP.DimensionI m → DSL (x ∷ I|i) m
+setX dsl = do
+  keys ← lastWithKeys ["pixels", "percents"] dsl
+  F.for_ (keys ∷ Array Foreign) $ set "x"
+
+setY ∷ ∀ i m. Monad m ⇒ DSL TP.DimensionI m → DSL (y ∷ I|i) m
+setY dsl = do
+  keys ← lastWithKeys ["pixels", "percents"] dsl
+  F.for_ (keys ∷ Array Foreign) $ set "y"
+
+setZ ∷ ∀ i m. Monad m ⇒ DSL TP.DimensionI m → DSL (z ∷ I|i) m
+setZ dsl = do
+  keys ← lastWithKeys ["pixels", "percents"] dsl
+  F.for_ (keys ∷ Array Foreign) $ set "z"

```

Comment:
I take it these don't work as:
``` purescript
F.traverse_ (set "z") =<< lastWithKeys ["pixels", "percents"] dsl
```
because you need the type annotation in the middle?

Reply:
Exactly. Having `Alternative` is kinda useful, but more annotations are needed

Reply:
Yeah, Haskell's defaulting is one solution to this, but it can be kinda weird/unexpected also, so this isn't too bad a price to pay :wink:

An interface with proxy support for some of these things is a nice option when we get `@` syntax: you could do provide functions for things like `fromFoldable <<< toFoldable' @List` to help with those cases.